core/test/base/value_typed_lin_op.cpp

```diff
@@ -0,0 +1,176 @@
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/lin_op.hpp>
+
+
+#include <complex>
+#include <memory>
+#include <type_traits>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/test/utils.hpp"
+
+
+namespace {
+
+
+template <typename T>
+T get_value_from_linop(const gko::LinOp *val);
+
+
+template <typename T>
+class DummyValueTypedLinOp
+    : public gko::EnableValueTypedLinOp<DummyValueTypedLinOp<T>, T>,
+      public gko::EnableCreateMethod<DummyValueTypedLinOp<T>> {
+    using Self = DummyValueTypedLinOp<T>;
+
+public:
+    explicit DummyValueTypedLinOp(std::shared_ptr<const gko::Executor> exec,
+                                  gko::dim<2> size = gko::dim<2>{})
+        : gko::EnableValueTypedLinOp<Self, T>(exec, size), value_()
+    {}
+    DummyValueTypedLinOp(std::shared_ptr<const gko::Executor> exec,
+                         gko::dim<2> size, T value)
+        : gko::EnableValueTypedLinOp<Self, T>(exec, size), value_(value)
+    {}
+
+    T get_value() const { return value_; }
+
+protected:
+    void apply_impl(const gko::LinOp *b, gko::LinOp *x) const override {}
+
+    void apply_impl(const gko::LinOp *alpha, const gko::LinOp *b,
+                    const gko::LinOp *beta, gko::LinOp *x) const override
+    {
+        T alpha_v = get_value_from_linop<T>(alpha);
+        T beta_v = get_value_from_linop<T>(beta);
+
+        gko::as<Self>(x)->value_ =
+            alpha_v * gko::as<Self>(this)->value_ * gko::as<Self>(b)->value_ +
+            beta_v * gko::as<Self>(x)->value_;
+    }
+
+    T value_;
+};
+
+
+template <typename T>
+T get_value_from_linop(const gko::LinOp *val)
+{
+    if (auto *dense = dynamic_cast<const gko::matrix::Dense<T> *>(val)) {
+        return dense->at(0, 0);
+    } else {
+        return (dynamic_cast<const DummyValueTypedLinOp<T> *>(val))
+            ->get_value();
+    }
+}
+
+
+template <typename T>
+class EnableValueTypedLinOp : public ::testing::Test {
+protected:
+    using dummy_type = DummyValueTypedLinOp<T>;
+    using value_type = T;
+
+    gko::dim<2> dim{1, 1};
+    std::shared_ptr<const gko::ReferenceExecutor> ref{
+        gko::ReferenceExecutor::create()};
+    std::unique_ptr<dummy_type> op{dummy_type::create(ref, dim, T{1.0})};
+    T alpha_v{2.0};
+    std::unique_ptr<dummy_type> alpha{dummy_type::create(ref, dim, alpha_v)};
+    T beta_v{3.0};
+    std::unique_ptr<dummy_type> beta{dummy_type::create(ref, dim, beta_v)};
+    std::unique_ptr<dummy_type> b{dummy_type::create(ref, dim, T{4.0})};
+    std::unique_ptr<dummy_type> x{dummy_type::create(ref, dim, T{5.0})};
+};
+
+
+TYPED_TEST_SUITE(EnableValueTypedLinOp, gko::test::ValueTypes);
+
+
+TYPED_TEST(EnableValueTypedLinOp, CanCallExtendedApplyImplLinopLinop)
+{
+    using value_type = typename TestFixture::value_type;
+    this->op->apply(lend(this->alpha), lend(this->b), lend(this->beta),
```

Comment:
this->op can not be LinOP, right?
it needs to be the explicit type or EnableValuedTypeLinOp<T> type